op-batcher/batcher/throttle_controller_test.go

```diff
@@ -0,0 +1,769 @@
+package batcher
+
+import (
+	"math"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/ethereum-optimism/optimism/op-batcher/config"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// noopLogger creates a logger that discards all output for testing
+func noopLogger() log.Logger {
+	return log.Root()
+}
+
+// TestStepController tests the step controller behavior
+func TestStepController(t *testing.T) {
+	strategy := NewStepStrategy(1000000, 5000, 21000, 130000)
+	controller := NewThrottleController(strategy)
+
+	tests := []struct {
+		name              string
+		pendingBytes      uint64
+		expectedIntensity float64
+		expectedMaxTxSize uint64
+		expectedMaxBlock  uint64
+	}{
+		{
+			name:              "below threshold",
+			pendingBytes:      500000,
+			expectedIntensity: 0.0,
+			expectedMaxTxSize: 0,
+			expectedMaxBlock:  130000,
+		},
+		{
+			name:              "at threshold",
+			pendingBytes:      1000000,
+			expectedIntensity: 0.0,
+			expectedMaxTxSize: 0,
+			expectedMaxBlock:  130000,
+		},
+		{
+			name:              "above threshold",
+			pendingBytes:      1500000,
+			expectedIntensity: 1.0,
+			expectedMaxTxSize: 5000,
+			expectedMaxBlock:  21000,
+		},
+		{
+			name:              "far above threshold",
+			pendingBytes:      3000000,
+			expectedIntensity: 1.0,
+			expectedMaxTxSize: 5000,
+			expectedMaxBlock:  21000,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			params := controller.Update(tt.pendingBytes, 0)
+
+			if params.Intensity != tt.expectedIntensity {
+				t.Errorf("expected intensity %f, got %f", tt.expectedIntensity, params.Intensity)
+			}
+			if params.MaxTxSize != tt.expectedMaxTxSize {
+				t.Errorf("expected MaxTxSize %d, got %d", tt.expectedMaxTxSize, params.MaxTxSize)
+			}
+			if params.MaxBlockSize != tt.expectedMaxBlock {
+				t.Errorf("expected MaxBlockSize %d, got %d", tt.expectedMaxBlock, params.MaxBlockSize)
+			}
+		})
+	}
+}
+
+// TestLinearController tests the linear controller behavior
+func TestLinearController(t *testing.T) {
+	strategy := NewLinearStrategy(1000000, 5000, 21000, 130000, 2, noopLogger())
+	controller := NewThrottleController(strategy)
+
+	tests := []struct {
+		name              string
+		pendingBytes      uint64
+		expectedIntensity float64
+		tolerance         float64
+	}{
+		{
+			name:              "below threshold",
+			pendingBytes:      500000,
+			expectedIntensity: 0.0,
+			tolerance:         0.001,
```

Comment:
You reuse `tolerance` a lot with the same value, so it should be a constant hoisted up to the file level IMO.